### ISSUE TITLE
Mirror Strategy slow down by adding GPUs

### ISSUE THREAD

**Issue by sakh251** [tensorflow/tensorflow#32172]
I am using the custom estimator with TFrecord. I am training VGG16. By increasing the number of GPUs the training time increase. All GPUs are in a single machine. GPUs utilization is about 100%. So it seems the input function can feed the data to GPUs well. I am using tf.layers.conv2d, tf.contrib.layers.flatten, tf.layers.dense to create the VGG Model. However, I have a large number of groups_Deps nodes in the computation graph. since all GPUs are in a single machine I am wondering why increasing number of GPUs lead to decreasing the global_Steps/Sec and also increasing the training time. 

![image](https://user-images.githubusercontent.com/17527773/64166782-195cf380-ce48-11e9-91d8-15ceade1db54.png)

**Comment by oanush** [comment]
@sakh251 ,
Can you share a simple and standalone code to reproduce to reproduce the issue reported here.Also mention the TensorFlow version being used.Thanks!

**Comment by isaprykin** [comment]
Hi!  This is odd.

Is it the same if you 1) use tf.distribute.MirroredStrategy instead of the contrib one? 2) use RunConfig.experimental_distribute.train_distribute instead of RunConfig.train_distribute?

If it's the same, could you include the full picture of the graph?

**Comment by sakh251** [comment]
Thanks for your answer. Yes it is strange

For 1 I can not use tf.distribute.MirroredStrategy. I got 'module' object has no attribute 'distribute' error. I am using the 1.11.0 version.
For 2 Also I have same kind of error. But in this case, I am not sure if I am using it correctly. Cloud you please send me an example for this.
I tried to attached complete graph. however, it is too large. I hope these can be helpful

![image](https://user-images.githubusercontent.com/17527773/64701243-b94df900-d4a8-11e9-8125-a3ebc62b3041.png)

![image](https://user-images.githubusercontent.com/17527773/64701288-cf5bb980-d4a8-11e9-96df-f37b37b46236.png)

**Comment by xxllp** [comment]
I  have the same problem

**Comment by guptapriya** [comment]
Hi, can you please try with latest version of TF and see if you still have the same problem?

**Comment by tensorflowbutler** [comment]
It has been 14 days with no activity and the `awaiting response` label was assigned. Is this still an issue?

**Comment by sakh251** [comment]
I still have this problem and could not solve it> I need some clue to track the problem

**Comment by oscaralvaro** [comment]
@guptapriya I see the same problem of slowness with MirroredStrategy today, using tf-nightly==2.4.0.dev20200701

Could we please get some more elaborated answer regarding what could be the problem? Some of us have invested time coding for using this assuming that it was production ready/tested and provided performance gains.

**Comment by guptapriya** [comment]
@oscaralvaro MirroredStrategy is certainly production ready and tested. Can you please open another bug with your use case? the current bug uses TF 1 and contrib layers etc which we are no longer supporting. 
Please provide code to repro, as well as use the profiler to get profiles during training and share the profiles. See these links for how to use the profiling:
https://www.tensorflow.org/guide/profiler
https://www.tensorflow.org/tensorboard/tensorboard_profiling_keras

**Comment by guptapriya** [comment]
cc @nikitamaia

**Comment by oscaralvaro** [comment]
@guptapriya thank you for your response. It is not trivial to isolate an example because I'm using it in conjunction with keras-tuner and preprocessing layers. I will try to create an example where I can observe the same slow effect when using MirroredStrategy. However, before that I wanted to know if there is any guidelines/documentation regarding the performance considerations when using TF with multi-gpu. So far in the documentation I only could find recommendations regarding the global_batch_size and learning rate tuning

**Comment by guptapriya** [comment]
@nikitamaia will be working on such a guide specifically for multi GPU. In the meantime, some of the common guidelines from here will also apply if input pipeline is the bottleneck, which can often be the case:
https://www.tensorflow.org/guide/data_performance

For using the profiler, you don't need to isolate the example. So perhaps you should try running the profiler. The profiler output often gives valuable suggestions like whether input pipeline is bottleneck or something else.

**Comment by oscaralvaro** [comment]
Thank you @guptapriya for the suggestions. I'm going to run the profiler first and see if I can fix the issue before spending time isolating the example. Looking forward for the performance in multi-gpu guide by @nikitamaia

**Comment by oscaralvaro** [comment]
@guptapriya I've created the issue https://github.com/tensorflow/tensorflow/issues/41213, with code to reproduce the behavior. In summary I observe the following difference in training time:

#average step for single-gpu
Epoch 2/50
8/8 [==============================] - 0s 5ms/step - loss: 0.7215 - accuracy: 0.5620 - val_loss: 0.6550 - val_accuracy: 0.7213

#average step for multi-gpu, is more than 10 times slower!!
Epoch 2/50
2/2 [==============================] - 0s 224ms/step - loss: 0.6756 - accuracy: 0.5702 - val_loss: 0.6435 - val_accuracy: 0.6885

**Comment by mohantym** [comment]
Hi  @sakh251 ! we are  checking to see if you are still looking for assistance in this issue.
Could you please try on latest stable version of  TF 2.6  and let us know if this is still an issue.Thanks!

**Comment by MoFHeka** [comment]
@sakh251 Hi, did you find the solution?

**Comment by mustafahalimeh** [comment]
I am also facing the same issue with MirroredStrategy in TF 2.6.2. 
2 GPUS is faster than using 4 GPUS for Cifar images dataset.
